### DISCUSSION TITLE
为什么数据库会丢失数据 · Why's THE Design? · /whys-the-design-database-data-loss

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#588] [Announcements]
https://draveness.me/whys-the-database-data-loss

**Comment by yuchanns** [comment]
请教一下延伸阅读的那些材料博主一般都是从哪里获取的？或者说，一般都是从哪些渠道获取相关资讯

**Comment by draveness** [comment]
> 请教一下延伸阅读的那些材料博主一般都是从哪里获取的？或者说，一般都是从哪些渠道获取相关资讯

Google 啊，不过先看 Google Scholar 然后看 Blog，遍历页面的所有链接

**Comment by liyebing** [comment]
道理我都懂，但我肯定写不出这样表述清晰，格式优美的好文章

**Comment by ray1888** [comment]
第三个问题要看是传统的服务器还是非传统的服务器，传统的如果是真实落盘，理论上fsync下去成功才算落盘（但是像现在的数据库有log的，只要log成功落盘，其实就可以算真实写入了），但是对于现在又类似于傲腾（非易失性内存） 的情况下，其实写到那里就应该可以算落盘成功了？但是这个要看数据库怎样去适配新的硬件？

**Comment by draveness** [comment]
> 第三个问题要看是传统的服务器还是非传统的服务器，传统的如果是真实落盘，理论上fsync下去成功才算落盘（但是像现在的数据库有log的，只要log成功落盘，其实就可以算真实写入了），但是对于现在又类似于傲腾（非易失性内存） 的情况下，其实写到那里就应该可以算落盘成功了？但是这个要看数据库怎样去适配新的硬件？

非易失性内存这个我之前还真没了解过，它与内存和磁盘之间有明显的界限么，区别是什么呢？

**Comment by ray1888** [comment]
它就是兼有内存和硬盘的特点，既能持久化存储，而且有近似于内存的读写速度，缺点就是贵和容量比较少，但是用来做cache是非常适合的

**Comment by xx1906** [comment]
Redis 的 RDB 和 AOF 机制什么时候会将数据落盘？
----
答：redis RDB, AOF 机制会有这个过程 
```c
// rdb.c/aof.c 文件中的代码， 
if (fflush(fp) == EOF)
    goto werr;
  // 将数据刷会磁盘
if (fsync(fileno(fp)) == -1)
    goto werr;
```
